core/src/main/scala/it/scalachess/core/logic/moves/generators/GeneratePawnMoves.scala

```diff
@@ -0,0 +1,104 @@
+package it.scalachess.core.logic.moves.generators
+
+import it.scalachess.core.{ Black, Color, White }
+import it.scalachess.core.board.{ Board, Position }
+import it.scalachess.core.logic.moves.{ ValidMove, ValidPromotion, ValidSimpleMove }
+import it.scalachess.core.pieces.{ Bishop, Knight, Pawn, Piece, PieceType, Queen, Rook }
+import scalaz.{ Failure, Success, Validation }
+
+private[generators] object GeneratePawnMoves extends GeneratePieceMoves {
+
+  override def apply(color: Color, board: Board, from: Position): List[ValidMove] =
+    color match {
+      case White => generatePawnSimpleMoves(Pawn, color, board, from, from.rowUpMod, Board.whitePawnsStartingRow)
+      case Black => generatePawnSimpleMoves(Pawn, color, board, from, from.rowDownMod, Board.blackPawnsStartingRow)
+    }
+
+  def generatePawnSimpleMoves(pieceType: PieceType,
+                              color: Color,
+                              board: Board,
+                              from: Position,
+                              forwardRowMod: Int,
+                              startingRow: Int): List[ValidMove] = {
+    val moveOnePosWithoutCapture =
+      generatePawnMovement(pieceType, color, board, from, Position.of(from.col, from.row + forwardRowMod))
+    val moveTwoPosWithoutCapture =
+      from.row match {
+        case `startingRow` =>
+          moveOnePosWithoutCapture match {
+            case Failure(errorMsg) => Failure(errorMsg)
+            case Success(_) =>
+              generatePawnMovement(pieceType,
+                                   color,
+                                   board,
+                                   from,
+                                   Position.of(from.col, from.row + forwardRowMod + forwardRowMod))
+          }
+        case _ =>
+          Failure("Pawn's movement: can't move two position forward, because it's not located in the starting row")
+      }
+    val leftAttack =
+      generatePawnAttack(pieceType,
+                         color,
+                         board,
+                         from,
+                         Position.of(from.col - from.colLeftMod, from.row + forwardRowMod))
+    val rightAttack =
+      generatePawnAttack(pieceType,
+                         color,
+                         board,
+                         from,
+                         Position.of(from.col + from.colRightMod, from.row + forwardRowMod))
+    List(moveOnePosWithoutCapture, moveTwoPosWithoutCapture, leftAttack, rightAttack)
+      .filter(_.toOption.nonEmpty)
+      .map(_.toOption.get)
+      .flatMap(move => {
+        if (isWhitePromoting(move) || isBlackPromoting(move))
```

Comment:
if (isPromoting(move))

private def isPromoting(move: ValidSimpleMove): Boolean = 
(move.color == White && move.to.row == 8) || (move.color == Black && move.to.row == 1)